### PR TITLE
Toni day 6

### PR DESCRIPTION
**Pull request by tonilopezmr**
# Toni's solution for day 6 challenge

## Description

Divide and rule has been my tactic for today 

## Solution

The main problem  here

```kotlin
private fun balance(input: List<Int>): List<Int> {
    var (index, bank) = chooseBank(input)
    val distribution = if (bank < input.size) bank else input.size - 1
    val blocksForEach = bank / distribution
    val list = input.toMutableList()
    val displacement = blocksForEach * distribution
    list[index] = bank - displacement
    index++
    for (i in 1..distribution) {
      index %= input.size
      list[index] = list[index] + blocksForEach
      index++
    }

    return list
  }
```

## Tests battery

> This section can be the most interesting. Here, you can show the tests you have written for checking your solution.
You should explain why and which parts you are testing and how.

Happy Advent of Code!

### CODE REVIEW

File: tonilopezmr/day6/MemoryBank.kt [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import kategory.combine
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+typealias P = Pair<Int, Int>
+
+class MemoryBank {
+
+  private fun chooseBank(input: List<Int>): Pair<Int, Int> =
+      input.foldRightIndexed(P(0, 0)) { idx, bank, (index, blocks) ->
```
Comment (Cotel) [inline tonilopezmr/day6/MemoryBank.kt:12]: Is there any difference by starting from the right? 🤔

File: tonilopezmr/day6/MemoryBank.kt [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    list[index] = bank - displacement
+    index++
+    for (i in 1..distribution) {
+      index %= input.size
+      list[index] = list[index] + blocksForEach
+      index++
+    }
+
+    return list
+  }
```
Comment (Cotel) [inline tonilopezmr/day6/MemoryBank.kt:31]: Simple enough, not accostumed to see that much mutability in Kotlin 😅

File: tonilopezmr/day6/MemoryBank.kt [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    return count
+  }
+
+  tailrec private fun infiniteLoopCircleIn(input: List<Int>, history: List<List<Int>> = emptyList(), count: Int = 0): Int =
+      if (history.contains(input)) count
+      else {
+        val newInput = balance(input)
+        infiniteLoopCircleIn(newInput, history.toMutableList().apply { add(input) }, count + 1)
+      }
```
Comment (Cotel) [inline tonilopezmr/day6/MemoryBank.kt:51]: Recursion is the easiest way to go for this 😉

**Review by Cotel — APPROVED** [review]
Good job

File: tonilopezmr/day6/MemoryBank.kt (hunk 2 above)
Comment (tonilopezmr) [reply tonilopezmr/day6/MemoryBank.kt:31]: xD that helps me understand the problem, you can see the "immutable" way also 👍

File: tonilopezmr/day6/MemoryBank.kt (hunk 3 above)
Comment (tonilopezmr) [reply tonilopezmr/day6/MemoryBank.kt:51]: but I don't know how to solve the history without mutability 🤔

**Pull request merged**